2017 Robot/src/org/usfirst/frc/team4999/lights/Packet.java

```diff
@@ -0,0 +1,162 @@
+package org.usfirst.frc.team4999.lights;
+
+import java.nio.ByteBuffer;
+
+/**
+ * A command sent to the arduino to light up a pixel or area of pixels with a certain color.
+ * Follows <a href="https://github.com/momentumfrc/2017Steamworks/wiki/Light-strings-on-the-Robot">this</a> structure.
+ * @author jordan
+ *
+ */
+public class Packet {
+	// The max size of a packet is 16 bytes
+	// Before a packet is sent, B_BUFFER is filled with that packet's data. B_BUFFER is then passed to the I2C writeBulk() method
+	private static final ByteBuffer B_BUFFER = ByteBuffer.allocateDirect(16);
```

Comment:
The ByteBuffer needs to be direct because if you look at the implementation of the I2C class, it requires a direct ByteBuffer, and throws an IllegalArgumentException() if the ByteBuffer is not direct.
I'm using ByteBuffer because if you look at the implementation of the I2C class, when you pass writeBulk() an array of bytes it creates a new direct ByteBuffer to hold them. I figured that it would be more memory efficient to just have one direct ByteBuffer that is read from rather than to be constantly creating new ByteBuffers, since direct buffers 'typically have somewhat higher allocation and deallocation costs than non-direct buffers' and it's recommended that 'direct buffers be allocated primarily for large, long-lived buffers'

Reply:
That makes a lot of sense when you explain it. I'm a little (a lot) disappointed in WPI for providing an API that's poorly documented in a way that could cause significant problems.

That said, since you have a good reason to use ByteBuffer, you should still be following some data encapsulation rules.  I would not have the singleton ByteBuffer be part of the Packet class, but part of the NeoPixel device class, or another class that wraps/extends the I2C base class.  That way, the details of Packet live only in Packet and the details of the I2C implementation live only there.

Reply:
Okay, I subclassed the I2C into NeoPixelIO, like you suggested earlier, so I'll make the buffer a field on that object